File: src/main/java/fi/quanfoxes/parser/Pattern.java
package fi.quanfoxes.parser;

import fi.quanfoxes.lexer.Token;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public abstract class Pattern {
    private ArrayList<Integer> path;

    public Pattern(Integer... path) {
        this.path = new ArrayList<>(Arrays.asList(path));
    }

    public abstract int priority(List<Token> tokens);
    public abstract boolean passes(List<Token> tokens);
    public abstract Node build(Context context, List<Token> tokens) throws Exception;

    public int start() {
        return 0;
    }

    public int end() {
        return -1;
    }

    public ArrayList<Integer> getPath() {
        return path;
    }
}

File: Vivid/Parser/Pattern.cs
using System.Collections.Generic;

public abstract class Pattern
{
	public List<int> Path { get; set; } = new();
	public int Priority { get; set; }
	public long Id { get; set; }
	public bool IsConsumable { get; set; } = true;

	public Pattern(params int[] path)
	{
		Path = new List<int>(path);
	}

	public abstract bool Passes(Context context, ParserState state, List<Token> tokens, int priority);
	public abstract Node? Build(Context context, ParserState state, List<Token> tokens);
}